Add tracing spans to CRI runtime service apis

swagatbora90 opened the pull request:
Rebased from main.

Most of the conflicts were resolved by replacing import from "github.com/containerd/containerd/tracing" with "github.com/containerd/containerd/v2/pkg/tracing". Also manually added the changes to internal/cri/instrument/instrumented_service.go` since the file location has now changed.

---
Rev 2

* Addressing this [comment](https://github.com/containerd/containerd/pull/7616#discussion_r1051198511) about adding additional spans at the instrumented service layer. Instead of creating a new span, we can simply use the span created at the gRPC layer and add additional attributes to it. Instrumented service is anyways a thin layer that does not require a new span of its own. Instead we capture and record the error in the parent span itself.

* Similar to Runtime APIs, we will only create spans in the mutating image service APIs. Removed span created for `ListImages`, `ImageStatus` and `ImageFSinfo` calls as they are non-mutating.


----
Rev 1

Add spans to CRI runtime service and related client methods

This adds otel spans to CRI service mainly targeting mutating apis which includes:
* Sandbox apis - RunPodSandbox, StopPodSandbox, RemovePodSandbox
* Container apis - CreateContainer, StartContainer, StopContainer, RemoveContainer
* Attach, Exec and Exec Sync
* Containerd client methods: container.go, client.go, process.go and task.go

A single otel trace is created with sub-spans per component, for example:

gRPC call span <----> CRI service span  <----> Containerd client method span <----> metadata span

Signed-off-by: Swagat Bora <sbora@amazon.com>

## Sandbox 
### RunPodSandbox
<img width="1728" alt="Screen Shot 2022-11-17 at 3 44 51 PM" src="https://user-images.githubusercontent.com/5525370/202585952-e050ccbe-7b7b-41aa-931c-22160b190810.png">

### StopPodSandbox
<img width="1565" alt="Screen Shot 2022-11-17 at 4 07 10 PM" src="https://user-images.githubusercontent.com/5525370/202586496-662b4a21-e942-4b9f-9e80-f79e7514b4bb.png">

### RemovePodSandbox
<img width="1728" alt="Screen Shot 2022-11-17 at 3 47 03 PM" src="https://user-images.githubusercontent.com/5525370/202585980-ca9094d3-8227-4140-b68c-54df710e8997.png">

## Container

### CreateContainer
<img width="1728" alt="Screen Shot 2022-11-17 at 3 47 31 PM" src="https://user-images.githubusercontent.com/5525370/202585998-fb034132-0a70-4827-b4df-3a872e798fc2.png">

### StartContainer
<img width="1728" alt="Screen Shot 2022-11-17 at 3 48 04 PM" src="https://user-images.githubusercontent.com/5525370/202586012-0cf59f2b-323a-4555-9cf3-1748fa52a259.png">

### StopContainer
<img width="1728" alt="Screen Shot 2022-11-17 at 3 48 30 PM" src="https://user-images.githubusercontent.com/5525370/202586028-6a9d8cb5-1ad2-4dc1-91cc-fd91b27c3397.png">

### RemoveContainer
<img width="1563" alt="Screen Shot 2022-11-17 at 4 09 06 PM" src="https://user-images.githubusercontent.com/5525370/202586837-c489e9b4-fa95-461e-8a32-c06446c7dde0.png">

swagatbora90 commented:
cc @kzys @henry118 @jterry75 @mxpv

kzys commented on container.go:
```diff
@@ -137,6 +138,11 @@ func (c *container) Labels(ctx context.Context) (map[string]string, error) {
 }
 
 func (c *container) SetLabels(ctx context.Context, labels map[string]string) (map[string]string, error) {
+	ctx, span := tracing.StartSpan(ctx, "container.SetLabels", tracing.SpanEventAttribute(
+		tracing.Attribute("container.id", c.id)),
+	)
+	defer span.End()
+	span.SetAttributes(tracing.Attribute("container.id", c.id))
```
Does it need to call both SpanEventAttribute and SetAttributes?

swagatbora90 replied:
No, removed `SetAttributes`

kzys commented on container.go:
```diff
@@ -317,6 +339,7 @@ func (c *container) Update(ctx context.Context, opts ...UpdateContainerOpts) err
 	if _, err := c.client.ContainerService().Update(ctx, r); err != nil {
 		return errdefs.FromGRPC(err)
 	}
+	span.AddEvent("container updated")
```
Do we need this event?

swagatbora90 replied:
hmm, now that I look at it, the span ending should be enough to know that the container is updated. Will remove this.

kzys commented on metadata/containers.go:
```diff
@@ -144,7 +151,9 @@ func (s *containerStore) Create(ctx context.Context, container containers.Contai
 		}
-
+		span.SetAttributes(
+			tracing.Attribute("container.CreatedAt", container.CreatedAt.Format(time.RFC3339)),
```
createdAt or CreatedAt?

swagatbora90 replied:
will change this to use smaller case like the rest.

kzys commented on pkg/cri/server/container_start.go:
```diff
@@ -178,6 +183,8 @@ func (c *criService) StartContainer(ctx context.Context, r *runtime.StartContain
 
 	containerStartTimer.WithValues(info.Runtime.Name).UpdateSince(start)
+	span.AddEvent("container started",
+		tracing.SpanEventAttribute(tracing.Attribute("container.start.duration", time.Since(start).String())))
```
Isn't it the same as the duration of the span?

swagatbora90 replied:
Since, I am using `tracing.SpanFromContext(ctx)`, I cannot be sure what the original span is coming from and hence the duration here might be useful.

kzys commented on pkg/cri/server/sandbox_run.go:
```diff
@@ -326,6 +334,9 @@ func (c *criService) RunPodSandbox(ctx context.Context, r *runtime.RunPodSandbox
 
 		sandboxCreateNetworkTimer.UpdateSince(netStart)
+		span.AddEvent("finished pod network setup",
+			tracing.SpanEventAttribute(tracing.Attribute("pod.network.setup.duration", time.Since(netStart).String())),
```
How does it access the current span?

swagatbora90 replied:
I am getting the span from the context in L64?

kzys commented on pkg/cri/server/sandbox_status.go:
```diff
@@ -37,7 +37,6 @@ func (c *criService) PodSandboxStatus(ctx context.Context, r *runtime.PodSandbox
 	if err != nil {
 		return nil, fmt.Errorf("an error occurred when try to find sandbox: %w", err)
 	}
-
```
Don't make unnecessary changes.

mxpv commented on tracing/tracing.go:
```diff
@@ -34,6 +34,14 @@ type StartConfig struct {
 type SpanOpt func(config *StartConfig)
 
+// WithSpanAttribute appends attributes to a new created span.
+func WithSpanAttribute(attributes ...attribute.KeyValue) SpanOpt {
```
1. Let's rename this from `WithSpanAttribute` to `WithAttribute`. This is going to be used with `tracing.StartSpan`, so it's already clear that the attributes are going to be applied to span and no need to repeat that.
2. Let's have it accept `k string, v interface{}` instead of `attributes ...attribute.KeyValue`. This way we'll have it more ergonomic (less nested function calls):


Before:
```go
	ctx, span := tracing.StartSpan(ctx, "task.Kill",
 		tracing.WithSpanAttribute(
 			tracing.Attribute("task.id", t.ID()),
 			tracing.Attribute("task.pid", int(t.Pid()))))
```

After:
```go
	ctx, span := tracing.StartSpan(ctx, "task.Kill",
 		tracing.WithAttribute("task.id", t.ID()),
 		tracing.WithAttribute("task.pid", int(t.Pid())),
        )
```

WDYT?

swagatbora90 replied:
Sounds good. I have made the change acordingly

mxpv commented on metadata/containers.go:
```diff
@@ -254,11 +269,13 @@ func (s *containerStore) Update(ctx context.Context, container containers.Contai
 func (s *containerStore) Delete(ctx context.Context, id string) error {
+	ctx, span := tracing.StartSpan(ctx, tracing.Name(spanContainerPrefix, "Delete"))
+	defer span.End()
+	span.SetAttributes(tracing.Attribute("container.id", id))
```
nit: `WithAttribute`?

(same for below)

mxpv commented on pkg/cri/server/sandbox_run.go:
```diff
@@ -326,6 +334,9 @@ func (c *criService) RunPodSandbox(ctx context.Context, r *runtime.RunPodSandbox
 
 		sandboxCreateNetworkTimer.UpdateSince(netStart)
+		span.AddEvent("finished pod network setup",
+			tracing.EventAttribute(tracing.Attribute("pod.network.setup.duration", time.Since(netStart).String())),
```
Should we update `AddEvent` to accept the list of attributes? (I haven't seen other options passed), so it'll be:


```go
span.AddEvent("finished pod network setup", tracing.Attribute("pod.network.setup.duration", time.Since(netStart).String())
```

WDYT?

swagatbora90 replied:
Yes, I kept this so that we can pass `trace.EventOption` (which includes attributes, timestamps and stacktrace).  For our current use case we are only setting attributes, so I am fine just accepting attributes.

mxpv commented on pkg/cri/server/sandbox_run.go:
```diff
@@ -386,6 +397,13 @@ func (c *criService) RunPodSandbox(ctx context.Context, r *runtime.RunPodSandbox
+	sandboxStatus := sandbox.Status.Get()
+	span.AddEvent("Successfully started sandbox container",
+		tracing.EventAttribute(
+			attribute.Int("sandbox.status.pid", int(sandboxStatus.Pid)),
```
`tracing.Attribute("sandbox.status.pid", int(sandboxStatus.Pid))`
will pass this as int attribute, right? (no need to have `attribute.Int` here).

mxpv commented on process.go:
```diff
@@ -27,6 +27,8 @@ import (
 	"github.com/containerd/containerd/errdefs"
 	"github.com/containerd/containerd/protobuf"
+	"github.com/containerd/containerd/tracing"
+	"go.opentelemetry.io/otel/attribute"
```
Let's not import opentelemetry everywhere (see my comment above).
We should probably use `tracing.Attribute` instead with proper types passed in.

mxpv commented on pkg/cri/server/container_execsync.go:
```diff
@@ -72,6 +73,9 @@ func (cw *cappedWriter) isFull() bool {
 // If command exits with a non-zero exit code, an error is returned.
 func (c *criService) ExecSync(ctx context.Context, r *runtime.ExecSyncRequest) (*runtime.ExecSyncResponse, error) {
+	span := tracing.SpanFromContext(ctx)
+	span.SetAttributes(tracing.Attribute("container.id", r.GetContainerId()))
```
Can `tracing.SpanFromContext()` also accept arbitrary attributes?

swagatbora90 replied:
I actually prefer `SpanFromContext(ctx)` to return the existing span as is from context. This way it is more readable and clear what each api method is used for.

mxpv reviewed (commented):
Mostly nit.

Also once LGTMed, we should carry this to `sbserver`.

estesp commented:
Looks like merge commits got added rather than a rebase in the last push?

swagatbora90 commented:
> Looks like merge commits got added rather than a rebase in the last push?

Fixing this

mikebrow commented on pkg/cri/server/instrumented_service.go:
```diff
@@ -232,13 +235,16 @@ func (in *instrumentedService) StopPodSandbox(ctx context.Context, r *runtime.St
 	if err := in.checkInitialized(); err != nil {
 		return nil, err
 	}
+	ctx, span := tracing.StartSpan(ctx, tracing.Name(criSpanPrefix, "StopPodSandbox"))
```
again .. not sure we need to start span in instrumented, and Stop()and stop()  these id's won't always match

swagatbora90 replied:
We create a new span in the instrumented service and then we reference the same span in Stop() and stop(). Since each attribute in a span has to be unique, otel only keeps the last updated value for a attribute. In our case `container.id` and `sandbox.id` cannot have multiple values on the same span.

Anyways I will make the change to use the full id wherever I have added the attribute

mikebrow commented on pkg/cri/server/sandbox_stop.go:
```diff
@@ -67,6 +73,10 @@ func (c *criService) stopPodSandbox(ctx context.Context, sandbox sandboxstore.Sa
 			return fmt.Errorf("failed to stop container %q: %w", container.ID, err)
 		}
+
+		span.AddEvent("stopped container in sandbox",
```
same here we should do the events once .. and probably inside the inner func after conversion to the full id

swagatbora90 replied:
ack

mikebrow reviewed (changes requested):
Hello thanks for the PR.. please see questions/comments

swagatbora90 commented on pkg/cri/server/container_attach.go:
```diff
@@ -31,6 +32,8 @@ import (
 // Attach prepares a streaming endpoint to attach to a running container, and returns the address.
 func (c *criService) Attach(ctx context.Context, r *runtime.AttachRequest) (*runtime.AttachResponse, error) {
+	span := tracing.SpanFromContext(ctx)
+	span.SetAttributes(tracing.Attribute("container.id", r.GetContainerId()))
```
Makes sense. Will change this to use the container id from the store. During my tests, the ID in the request and the one we get from the store were the same, but I didn't realize that we can use partial id.

swagatbora90 commented on pkg/cri/server/container_stop.go:
```diff
@@ -29,13 +29,16 @@ import (
 // StopContainer stops a running container with a grace period (i.e., timeout).
 func (c *criService) StopContainer(ctx context.Context, r *runtime.StopContainerRequest) (*runtime.StopContainerResponse, error) {
+	span := tracing.SpanFromContext(ctx)
+	span.SetAttributes(tracing.Attribute("container.id", r.GetContainerId()))
```
ack

swagatbora90 commented on pkg/cri/server/container_stop.go:
```diff
@@ -66,6 +69,10 @@ func (c *criService) StopContainer(ctx context.Context, r *runtime.StopContainer
 
 	containerStopTimer.WithValues(i.Runtime.Name).UpdateSince(start)
 
+	span.AddEvent("container stopped",
```
Yes, tracing calls to NRI plugins is not in the scope currently. I will move this event to `stopContainer()`

Actually, we may be able to trace the calls on the NRI ttrpc layer. Otel provides automatic instrumentation of grpc calls but I don't think there is anything available to trace on the ttrpc layer. I found this [PR](https://github.com/open-telemetry/opentelemetry-go-contrib/pull/504) which adds grpc style interceptors for ttrpc but it never got merged.

mikebrow replied:
nod ... tracing into the ttrpc is a must for NRI and more importantly for our ttrpc shims..

swagatbora90 commented on pkg/cri/server/sandbox_stop.go:
```diff
@@ -46,10 +49,13 @@ func (c *criService) StopPodSandbox(ctx context.Context, r *runtime.StopPodSandb
 }
 
 func (c *criService) stopPodSandbox(ctx context.Context, sandbox sandboxstore.Sandbox) error {
+	span := tracing.SpanFromContext(ctx)
```
`tracing.SpanFromContext(ctx)` does not create a new span, only gets the existing span from context. In this case (and other CRI methods as well) a new span is created in the Instrumented service layer using `tracing.StartSpan()`

```
func (in *instrumentedService) StopPodSandbox(ctx context.Context, r *runtime.StopPodSandboxRequest) (_ *runtime.StopPodSandboxResponse, err error) {
	if err := in.checkInitialized(); err != nil {
		return nil, err
	}
	ctx, span := tracing.StartSpan(ctx, tracing.Name(criSpanPrefix, "StopPodSandbox"))
	defer span.End()
```

So, in `StopPodSandbox` and `stopPodSandbox` we are just adding additional attributes/events to that span as we traverse through internal methods. We create a new span again when the call reaches containerd client from CRI.

mikebrow replied:
we should talk.. the cri plugin is built in and uses the containerd services directly .. it's not remote any more.. A lot of this tracing of very thin internal layers will not be useful information.

kiashok commented:
@swagatbora90 Looks like all the PR comments were addressed. Do we know if anything is blocking sign offs? Was trying to add some logs to these APIs and came across this PR which will be useful to have!

swagatbora90 commented:
> @swagatbora90 Looks like all the PR comments were addressed. Do we know if anything is blocking sign offs? Was trying to add some logs to these APIs and came across this PR which will be useful to have!

@kiashok This PR needs rebase as there has been some changes to the instrumented service. Let me send a new revision. One of the things suggested by @mikebrow was to use the parent span created at the gRPC layer and not create a new span again at the instrumented service layer. I was planning to add that as a follow up PR, but I will go ahead and just make the change here itself. 

We do have to add similar changes to the new sandbox service, which I plan to add once this PR gets approved.

henry118 commented:
@swagatbora90 is still working on this. Reopen this PR to continue to work.

swagatbora90 commented:
Rebased from `main`.

Most of  the conflicts were resolved  by replacing import from  `"github.com/containerd/containerd/tracing"` with `"github.com/containerd/containerd/v2/pkg/tracing"`. Also manually added the changes to internal/cri/instrument/instrumented_service.go` since the file location has now changed.

estesp commented:
Thanks for getting this rebased! One of the commits has at least one whitespace error; you can see the exact issues with this command:
```
git show --check 54d64ac12e0c3b1872e7ead3232e7ba2f3d75515
```

estesp commented:
/ok-to-test

swagatbora90 commented:
> Thanks for getting this rebased! One of the commits has at least one whitespace error; you can see the exact issues with this command:
> 
> ```
> git show --check 54d64ac12e0c3b1872e7ead3232e7ba2f3d75515
> ```

`git-validation` is now passing
```
git-validation -range HEAD~2..HEAD -run DCO,short-subject,dangling-whitespace
INFO[0000] using commit range: HEAD~2..HEAD             
 * 00a3c7827 "Remove extra span created in the instrumented service layer" ... PASS
 * a26127478 "Add spans to CRI runtime service and related client methods" ... PASS`
 ```

estesp commented:
@mikebrow and @kzys: this PR has been updated and rebased and responded to all prior feedback; if you still are requesting changes can you comment ASAP; or dismiss your stale review if not. Thanks!

mikebrow reviewed (approved):
LGTM on rebase.. cheers

mikebrow reviewed (approved):
LGTM

kzys reviewed (approved):
Sorry for the late review. Approved.
